tests/apollo/test_skvbc_preexecution.py

```diff
@@ -191,3 +191,31 @@ async def test_view_change(self, bft_network):
                 break
 
         self.assertEqual(new_last_block, last_block)
+
+    @with_trio
+    @with_bft_network(start_replica_cmd)
+    async def test_f_nonprimary_crash_amid_parallel_tx(self, bft_network):
+        '''
+        Crash f nonprimary replicas and submit X parallel write submissions.
+        Block processing of the network should be unaffected with f-count interruption.
+        Final block length should match submitted transactions count exactly.
+        '''
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+        
+        num_of_requests = 50
+        client = bft_network.random_client()
+        nonprimaries = bft_network.all_replicas(without={0}) # primary index is 0
+        crash_targets = random.sample(nonprimaries, bft_network.config.f) # pick random f to crash
+        bft_network.stop_replicas(crash_targets) # crash chosen replicas
+        
+        started_block_count = skvbc.parse_reply(await client.read(skvbc.get_last_block_req()))
+        await self.run_concurrent_pre_execution_requests(skvbc, [client], num_of_requests, write_weight=1.0)
```

Comment:
By the way, just noticed you're passing  a single client here. This means effectively the requests will execute sequentially. This explains why the test is so slow by the way...

Reply:
PS: pre-execution is all about parallel request execution!

Reply:
Oh, is that so. I'll update to include many clients. What would be a reasonable concurrency level for this particular test..?

Reply:
Let's try with MAX_CONCURRENCY (10). On the other hand, due to batching, some requests might get executed within a single block. So you're not guaranteed to have as many blocks as individual requests.

Reply:
Yes, tests are now failing intermittently. Block counts don't line up.

Instead of counting blocks, is there a way to count writes?

Reply:
Yep, use the "lastExecutedSeqNum" metric

Reply:
Here's an example how to read it from a given replica:
https://github.com/vmware/concord-bft/blob/028883b383e2ba5cb3e68708e2c9d2ea6ab2b4b2/tests/apollo/test_skvbc_ro_replica.py#L158

Reply:
How about using `await bft_network.wait_for_last_executed_seq_num` function?

Reply:
Yep, sounds good too!

Reply:
I don't think batching is the problem.

I think the problem is `btf_network.random_clients(MAX_CONCURRENCY)` not returning 10 clients as it should. I don't know why it's not returning all 10 as it should.

Reply:
As long as `num_requests` is an exact multiple of submitter `clients`, checking the test by block creation alone totally works.

Reply:
I see Yulia skipped `test_sequential_pre_process_requests` and `test_concurrent_pre_process_requests` tests saying it's unstable. But after the `run_concurrent_pre_execution_requests` fix and giving 2~3 seconds for SKVBC to initialize. All tests pass.

Reply:
I'm observing a lot of intermittent `TooSlow` errors if there is no initialization grace time after `skvbc = kvbc.SimpleKVBCProtocol(bft_network)`

Reply:
Let's focus on our test here, ignore the unstable tests pls, Yulia is working on them. I suggest the following:
- add a wait after SKVBC replicas startup
- use 10 clients for the concurrent execution
- check for last executed sequence, instead of block number

Reply:
`lastExecutedSeqNum` is always returning 1/4 of final block number.

Reply:
Yes, it's entirely possible, due to batching, as I told you previously...

Reply:
OK, let's stick with block number, as long as the test is stable.

Reply:
I've tried querying all available nodes after waiting 60 seconds.
```
Replica ID 1, last executed sequence num = 4
Replica ID 2, last executed sequence num = 4
Replica ID 5, last executed sequence num = 4
Replica ID 6, last executed sequence num = 4
Randomly picked replica indexes [4, 3] (nonprimary) to be stopped.
Total of 18 write pre-exec tx, concurrently submitted through 9 clients.
Finished at block 18 (expected >= 18).
```
Whereas `lastExecutedSeqNum` is not exactly 1/4 of final block number, it does not seem to reflect number of actually submitted writes. [Concord metrics](https://confluence.eng.vmware.com/display/BLOC/Concord+Metrics) states that `lastExecutedSeqNum` is actually consensus counter, so doesn't it mean that batching works the other way...? e.g. Multiple block writes per single consensus (multiple blocks per single sequence number).

Reply:
Sorry, I mislead you yesterday... Batching groups client requests together in what we call a "decision block" with a sequence number. This decision block goes through consensus as a unit and also executes as a unit. Executing this decision block results in an actual block in the chain. This is what the lastExecutedSeqNum metric actually counts. 

So we should have as many blocks as executed sequence numbers. However, we may have fewer blocks than the number of client requests as they get grouped together before they go through consensus.

I hope it's clearer now...